rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,367 @@
+| S0  | Start          | The upload is never actually in this state.                                                 |
+| S1  | Waiting        | The upload is enqueued, and waiting for computing resources.                                |
+| S2  | Uploading      | The file is actively being uploaded.                                                        |
+| S3  | Uploaded       | The upload was completed successfully.                                                      |
+| S4  | Failed         | The upload has failed. A new link or file must be uploaded. Any upload progress is deleted. |
+| S5  | Cancel Pending | The upload is in the process of being canceled                                              |
+| S6  | Canceled       | The upload has been canceled. Any upload progress is deleted.                               |
+
+Once the upload is complete validation can begin.
+
+##### S0 - >S1
+
+An upload moves from S0 to S1 when an upload request has been made, and the system is waiting for resources to begin uploading.
+
+##### S1 - >S2
+
+An upload moves from S1 to S2 when the file is actively being uploaded to S3.
+
+##### S2 -> S5
+
+Once a user requests an upload to be canceled the state changes to Cancel Pending. The cancellation may not happen
+immediately, therefore the status of the upload is changed to cancel pending until the upload has been canceled by the
+compute resource processing the job.
+
+##### S5 -> S6
+
+Once the upload has been canceled by the compute resource the state will be canceled.
+
+##### S2 -> S3
+
+The upload has been completed. The state of the upload is uploaded.
+
+##### S2 -> S1
+
+If the upload fails, the job returns to the queue to be retried at a later time. The state changes to waiting.
+
+#### S2 -> S4
+
+The upload job has failed to download after several retries. The state of the upload is failed.
+
+### Test plan
+
+1. Verify that a large file can be uploaded using a shared link from Dropbox.
+1. Verify that a small file can be uploaded using a shared link from Dropbox.
+1. Verify the upload queue is not cleared until the file is store in S3.
+1. Verify a failed upload is marked as failed.
+1. Verify an upload is retried if it fails.
+1. Verify only the owner can query the status of an upload.
+1. Verify only the owner can upload toa dataset to their collection.
+1. Verify partial uploads are erased from S3 if they fail or are canceled.
+1. Verify a user can start another upload if the original upload failed, or was canceled.
+1. Verify a shared folder from DropBox cannot be uploaded.
+1. Verify uploaded files are removed from the upload bucket, once they moved to the primary bucket.
+1. Verify an uploaded file can be overwritten.
+1. Verify only [accepted file types](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations)
+   are allowed.
+
```

Comment:
I think it is also work verifying that we can handle multiple uploads simultaneously.    Start one large file upload, then immediately start a second independent large file upload.  Verify that the second uploads does not have to wait for the first one to finish, but rather a new upload server is started to handle the second request.

Reply:
added an additional test to verify multiple uploads